crates/ty_python_semantic/src/types/diagnostic.rs

```diff
@@ -2625,6 +2625,12 @@ pub(crate) fn report_undeclared_protocol_member(
                 SubclassOfInner::Dynamic(_) => return false,
             },
             Type::NominalInstance(instance) => instance.class(db),
+            Type::Union(union) => {
+                return union
+                    .elements(db)
+                    .iter()
+                    .all(|elem| should_give_hint(db, *elem));
+            }
```

Comment:
Admittedly more of a showcase for the refactor here. I'm also happy to remove this again (and the test above).